core/rawdb/chain_iterator.go

```diff
@@ -0,0 +1,227 @@
+					if n < int64(from) {
+						return
+					}
+				}
+				block := ReadBlock(db, ReadCanonicalHash(db, uint64(n)), uint64(n))
+				if prepare != nil && block != nil {
+					prepare(block)
+				}
+				// Feed the block to the aggregator, or abort on interrupt
+				select {
+				case results <- block:
+				case <-abort:
+					return
+				}
+			}
+		}()
+	}
+	// Reassemble the blocks into a contiguous stream and apply the action callback
+	var (
+		next, first, last int64
+		queue             = prque.New(nil)
+
+		batch  = db.NewBatch()
+		start  = time.Now()
+		logged = start.Add(-7 * time.Second) // Unindex during import is fast, don't double log
+	)
+	if !reverse {
+		next, first, last = int64(from), int64(from), int64(to)
+	} else {
+		next, first, last = int64(to-1), int64(to-1), int64(from-1)
+	}
+	for i := from; i < to; i++ {
+		// Retrieve the next result and bail if it's nil
+		block := <-results
+		if block == nil {
+			return errors.New("broken database")
+		}
+		// Push the block into the import queue and process contiguous ranges
+		priority := -int64(block.NumberU64())
+		if reverse {
+			priority = int64(block.NumberU64())
+		}
+		queue.Push(block, priority)
+		for !queue.Empty() {
+			// If the next available item is gapped, return
+			if _, priority := queue.Peek(); !reverse && -priority != next || reverse && priority != next {
+				break
+			}
+			// Next block available, pop it off and index it
+			block = queue.PopItem().(*types.Block)
+
+			if !reverse {
+				next++
+			} else {
+				next--
+			}
+			// Invoke action to inject specified data into key-value database.
+			action(batch, block)
+
+			// If enough data was accumulated in memory or we're at the last block, dump to disk
+			if batch.ValueSize() > ethdb.IdealBatchSize || next == last {
+				if err := batch.Write(); err != nil {
+					return err
+				}
+				batch.Reset()
+			}
+			// If we've spent too much time already, notify the user of what we're doing
+			if time.Since(logged) > 8*time.Second {
+				log.Info(progMsg, "blocks", int64(math.Abs(float64(next-first))), "total", to-from, "tail", block.Number(), "hash", block.Hash(), "elapsed", common.PrettyDuration(time.Since(start)))
+				logged = time.Now()
+			}
+		}
+	}
+	tail := to
+	if reverse {
+		tail = from
+	}
+	log.Info(doneMsg, "blocks", to-from, "tail", tail, "elapsed", common.PrettyDuration(time.Since(start)))
+	return nil
+}
+
+// InitDatabaseFromFreezer reinitializes an empty database from a previous batch
+// of frozen ancient blocks. The method iterates over all the frozen blocks and
+// injects into the database the block hash->number mappings.
+func InitDatabaseFromFreezer(db ethdb.Database) {
+	// If we can't access the freezer or it's empty, abort
+	frozen, err := db.Ancients()
+	if err != nil || frozen == 0 {
+		return
+	}
+	// hashBlock calculates block hash in advance using the multi-routine's concurrent
+	// computing power.
+	hashBlock := func(block *types.Block) { block.Hash() }
```

Comment:
This somehow seems odd to me... I mean, later on, during iteration, we _will_ look up the hash _from ancients_, and then load the header and body. So it feels weird that we need to hash it again?

All we theoretically need to do is,
 
* Traverse `ancient`s hashes, call `WriteHeaderNumber(batch, hash, number)` for each hash. There's no intrinsic need to actually go and dig up the (potentailly large) body.

Reply:
True. I guess it was just a generalization of the iteration mechanism to use blocks and not just headers. There's indeed no reason to dig up the bodies (lol) if we only need the heads (lol).

Reply:
Here's how I would do it
```diff

diff --git a/core/rawdb/chain_iterator.go b/core/rawdb/chain_iterator.go
index 7dfe69d70f..668ecc2dfc 100644
--- a/core/rawdb/chain_iterator.go
+++ b/core/rawdb/chain_iterator.go
@@ -166,19 +166,41 @@ func InitDatabaseFromFreezer(db ethdb.Database) {
        if err != nil || frozen == 0 {
                return
        }
-       // hashBlock calculates block hash in advance using the multi-routine's concurrent
-       // computing power.
-       hashBlock := func(block *types.Block) { block.Hash() }
-
-       // writeIndex injects hash <-> number mapping into the database.
-       writeIndex := func(batch ethdb.Batch, block *types.Block) { WriteHeaderNumber(batch, block.Hash(), block.NumberU64()) }
+       var (
+               batch  = db.NewBatch()
+               start  = time.Now()
+               logged = start.Add(-7 * time.Second) // Unindex during import is fast, don't double log
+               hash   common.Hash
+       )
+       for i := uint64(0); i < frozen; i++ {
+               if h, err := db.Ancient(freezerHashTable, i); err != nil {
+                       log.Crit("Failed to init database from freezer", "err", err)
+               } else {
+                       hash = common.BytesToHash(h)
+               }
 
-       if err := iterateCanonicalChain(db, 0, frozen, hashBlock, writeIndex, false, "Initializing database from freezer", "Initialized database from freezer"); err != nil {
-               log.Crit("Failed to init database from freezer", "err", err)
+               WriteHeaderNumber(batch, hash, i)
+               // If enough data was accumulated in memory or we're at the last block, dump to disk
+               if batch.ValueSize() > ethdb.IdealBatchSize {
+                       if err := batch.Write(); err != nil {
+                               log.Crit("Failed to write data to db", "err", err)
+                       }
+                       batch.Reset()
+               }
+               // If we've spent too much time already, notify the user of what we're doing
+               if time.Since(logged) > 8*time.Second {
+                       log.Info("Initializing database from freezer", "blocks", i, "total", frozen, "tail", i, "hash", hash, "elapsed", common.PrettyDuration(time.Since(start)))
+                       logged = time.Now()
+               }
+       }
+       if err := batch.Write(); err != nil {
+               log.Crit("Failed to write data to db", "err", err)
        }
-       hash := ReadCanonicalHash(db, frozen-1)
+       batch.Reset()
+
        WriteHeadHeaderHash(db, hash)
        WriteHeadFastBlockHash(db, hash)
+       log.Info("Initialized database from freezer", "blocks", frozen, "tail", frozen, "elapsed", common.PrettyDuration(time.Since(start)))
 }
 
 // IndexTransactions creates txlookup indices of the specified block range.
```

Reply:
And skip the generalized version, at least for InitDatabaseFromFreezer. And then I'd do optimized versions of tx indexing, and only then would I look to see if they can be generalized, but I think the current generalization is too cumbersome

Reply:
Here's the PR for that: https://github.com/karalabe/go-ethereum/pull/20